File: Week3/code/Vectorize2.R
#!/usr/bin/env R

# Runs the stochastic Ricker equation with gaussian fluctuations

rm(list=ls())

stochrick<-function(p0 = runif(1000, 0.5, 1.5), r = 1.2, K = 1, sigma = 0.2, numyears = 100)
{
  #initialize
  N<-matrix(NA,numyears,length(p0))
  N[1,]<-p0
  
  for (pop in 1:length(p0)){#loop through the populations
    
    for (yr in 2:numyears){ #for each pop, loop through the years

      N[yr,pop] <- N[yr-1,pop] * exp(r * (1 - N[yr - 1,pop] / K) + rnorm(1,0,sigma))
    
    }
  
  }
 return(N)

}

# same function as above, but with improved performance

stochrickvect<-function(p0=runif(1000,.5,1.5),r=1.2,K=1,sigma=0.2,numyears=100)
{
  #initialize
  N<-matrix(NA,numyears,length(p0))
  N[1,]<-p0
    
  for (yr in 2:numyears){ #for each pop, loop through the years
    N[yr,] <- N[yr-1,] * exp(r * (1 - N[yr - 1,] / K) + rnorm(1,0,sigma)) # taking the whole row into consideration instead of row element by element
  }
 
  return(N)

}

# comparing the computational speeds of the two functions
s1 = system.time(res1<-stochrick())[3] # store elapsed time of stochrick() function in a variable s1
print(paste("Non-vectorized Stochastic Ricker takes", as.numeric(format(round(s1, 3), nsmall = 2)), "seconds."), quote = FALSE)

s2 = system.time(res2<-stochrickvect())[3] # store elapsed time of stochrickvect() function in variable s2
print(paste("Vectorized Stochastic Ricker takes", as.numeric(format(round(s2, 3), nsmall = 2)), "seconds."), quote = FALSE)

ratio = s1/s2 # calculates the ratio of the elapsed times of the two functions
ratio = as.numeric(format(round(ratio, 3), nsmall = 2)) # formatting the ratio value to output only two digits after the decimal point
print(paste("Vectorizing the function reduced its speed by a magnitude of", ratio), quote = FALSE)

# saving time values to be used later in the run_Vectorize.sh bash script
write.table(c(as.numeric(format(round(s1, 3), nsmall = 2)), as.numeric(format(round(s2, 3), nsmall = 2))), "../data/Vectorize2_R.txt", row.names = FALSE, col.names = FALSE)

File: Week3/code/Vectorize2.py
#!/usr/bin/env python3

"""A python version of Vectorize2.R"""

__appname__ = 'Vectorize2.py'
__author__ = 'Hovig Artinian (ha819@imperial.ac.uk)'
__version__ = '0.0.1'
__license__ = 'None'

## imports ##
import scipy as sc
import numpy as np
import time
import sys

## functions ##
def stochrick (p0 = np.random.uniform(0.5,1.5,1000), r = 1.2, K = 1, sigma = 0.2, numyears = 100):
    """Runs a simulation of a stochastic Ricker model using nested for loops"""

    #initialize
    N = sc.zeros((numyears, len(p0)))
    N[0,:] = p0

    for pop in range(len(p0)): # loop through the populations
        for yr in np.arange(numyears-1)+1: # loop through the years
            N[yr,pop] = N[yr-1, pop] * np.exp(r * (1 - N[yr-1, pop] / K) + np.random.normal(0, sigma, 1))

    return(N)

def stochrickvect (p0 = np.random.uniform(0.5,1.5,1000), r = 1.2, K = 1, sigma = 0.2, numyears = 100):
    """Runs a simulation of a stochastic Ricker model using vectorization"""
    
    #initialize
    N = sc.zeros((numyears, len(p0)))
    N[0,:] = p0

    for yr in np.arange(numyears-1)+1:
        N[yr,:] = N[yr-1,:] * np.exp(r * (1 - N[yr-1,:] / K) + np.random.normal(0, sigma, 1)) # taking the whole row into consideration instead of row element by element

    return(N)

## main function ##
def main(argv):
    """The main function compares the computational speed of the stochrick 
    and stochrickvect functions."""

    # comparing the computational speeds of the two functions
    print("Using loops, the time taken is:")
    start_time1 = time.time()
    stochrick()
    elapsed_time1 = time.time() - start_time1
    print(round(elapsed_time1, 3))

    print("Using the in-built vectorized function, the time taken is:")
    start_time2 = time.time()
    stochrickvect()
    elapsed_time2 = time.time() - start_time2
    print(round(elapsed_time2, 3))

    # saving time values to be used later in the run_Vectorize.sh bash script
    f = open("../data/Vectorize2_py.txt", "w")
    f.write(str(round(elapsed_time1, 3)))
    f.write("\n")
    f.write(str(round(elapsed_time2, 3)))
    f.write("\n")
    f.close()

    return 0

if (__name__ == "__main__"):
    status = main(sys.argv)
    sys.exit(status)
